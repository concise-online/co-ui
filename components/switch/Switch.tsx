import React, { HTMLAttributes } from 'react';
import classNames from 'classnames';
import Icon from '../icon';
import { runCallback } from '../_utils/function';
import { getPrefixCls } from '../_utils/config';

const prefixCls = getPrefixCls('switch');

interface Props extends Omit<HTMLAttributes<HTMLInputElement>, 'onChange'> {
  value?: boolean;
  size?: 'large' | 'small';
  disabled?: boolean;
  loading?: boolean;
  className?: string;
  onChange?: (value: boolean) => void;
}

function Switch(props: Props) {
  const { value: checked, size, disabled, loading, style, onChange, className, ...otherProps } = props;
  const classes = classNames(prefixCls, {
    [`${prefixCls}-disabled`]: disabled,
    [`${prefixCls}-active`]: checked,
    [`${prefixCls}-${size}`]: !!size,
    [`${prefixCls}-loading`]: loading,
  }, className);

  return (
    <label
      aria-checked={checked}
      className={classes}
      style={style}
    >
      <input
        {...otherProps}
        type="checkbox"
        disabled={disabled || loading}
        checked={checked}
        onChange={(e) => runCallback(onChange, e.target.checked, e)}
      />
      <span className={`${prefixCls}-dot`}>
        {loading && <Icon name="loading" className={`${prefixCls}-spin`} spin={0.8} />}
      </span>
    </label>
  );
}

export default Switch;
