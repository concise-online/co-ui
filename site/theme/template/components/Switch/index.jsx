import React from 'react';
import classNames from 'classnames';
import styles from './styles.module.less';

function Switch({ opened, className }) {
  return (
    <i
      className={classNames(styles.switch, {
        [styles.opened]: opened,
      }, className)}
    />
  );
}

export default Switch;
