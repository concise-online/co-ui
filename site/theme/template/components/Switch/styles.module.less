.switch {
  position: relative;
  display: inline-block;
  width: 14px;
  height: 2px;
  background: #333;
  transition: background .3s;
  vertical-align: middle;

  &:before {
    position: absolute;
    top: -5px;
    display: block;
    content: '';
    width: 100%;
    height: 100%;
    background: #333;
    transition: transform .3s;
  }

  &:after {
    position: absolute;
    top: 5px;
    display: block;
    content: '';
    width: 100%;
    height: 100%;
    background: #333;
    transition: transform .3s;
  }

  &.opened {
    background: transparent;

    &:before {
      transform: translateY(5px) rotate(45deg);
    }

    &:after {
      transform: translateY(-5px) rotate(-45deg);
    }
  }
}