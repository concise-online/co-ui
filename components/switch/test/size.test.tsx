import React from 'react';
import { render } from 'enzyme';
import Switch from '../index';

describe('switch color feature', () => {
  it('renders correctly', () => {
    const wrapper = render((
      <div>
        <Switch size="large" />
        <Switch />
        <Switch size="small" />
      </div>
    ));

    expect(wrapper).toMatchSnapshot();
  });
});
