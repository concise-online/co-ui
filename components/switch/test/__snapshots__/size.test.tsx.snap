// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`switch color feature renders correctly 1`] = `
<div>
  <label
    class="co-switch co-switch-large"
  >
    <input
      type="checkbox"
    />
    <span
      class="co-switch-element"
    >
      <span
        class="co-switch-dot"
      />
    </span>
  </label>
  <label
    class="co-switch"
  >
    <input
      type="checkbox"
    />
    <span
      class="co-switch-element"
    >
      <span
        class="co-switch-dot"
      />
    </span>
  </label>
  <label
    class="co-switch co-switch-small"
  >
    <input
      type="checkbox"
    />
    <span
      class="co-switch-element"
    >
      <span
        class="co-switch-dot"
      />
    </span>
  </label>
</div>
`;
